modules/src/applications/ics20_fungible_token_transfer/context.rs

```diff
@@ -1,5 +1,326 @@
+use core::str::FromStr;
+
+use sha2::{Digest, Sha256};
+use subtle_encoding::hex;
+
+use super::error::Error as Ics20Error;
+use crate::applications::ics20_fungible_token_transfer::acknowledgement::Acknowledgement;
+use crate::applications::ics20_fungible_token_transfer::packet::PacketData;
+use crate::applications::ics20_fungible_token_transfer::relay_application_logic::on_ack_packet::process_ack_packet;
+use crate::applications::ics20_fungible_token_transfer::relay_application_logic::on_recv_packet::process_recv_packet;
+use crate::applications::ics20_fungible_token_transfer::{
+    DenomTrace, HashedDenom, IbcCoin, VERSION,
+};
+use crate::applications::ics20_fungible_token_transfer::events::{AckEvent, AckStatusEvent, RecvEvent, TimeoutEvent};
+use crate::applications::ics20_fungible_token_transfer::relay_application_logic::on_timeout_packet::process_timeout_packet;
+use crate::core::ics04_channel::channel::{Counterparty, Order};
 use crate::core::ics04_channel::context::{ChannelKeeper, ChannelReader};
+use crate::core::ics04_channel::msgs::acknowledgement::Acknowledgement as GenericAcknowledgement;
+use crate::core::ics04_channel::packet::Packet;
+use crate::core::ics04_channel::Version;
+use crate::core::ics05_port::context::{PortKeeper, PortReader};
+use crate::core::ics24_host::identifier::{ChannelId, ConnectionId, PortId};
+use crate::core::ics26_routing::context::{ModuleOutputBuilder, OnRecvPacketAck};
+use crate::prelude::*;
+use crate::signer::Signer;
+
+pub trait Ics20Keeper:
+    ChannelKeeper + PortKeeper + BankKeeper<AccountId = <Self as Ics20Keeper>::AccountId>
+{
+    type AccountId: Into<String>;
+
+    /// Sets a new {trace hash -> denom trace} pair to the store.
+    fn set_denom_trace(&mut self, denom_trace: &DenomTrace) -> Result<(), Ics20Error>;
+}
+
+pub trait Ics20Reader:
+    BankReader<AccountId = <Self as Ics20Reader>::AccountId>
+    + AccountReader<AccountId = <Self as Ics20Reader>::AccountId>
+    + ChannelReader
+    + PortReader
+{
+    type AccountId: Into<String> + FromStr;
+
+    /// get_port returns the portID for the transfer module.
+    fn get_port(&self) -> Result<PortId, Ics20Error>;
+
+    /// Returns the escrow account id for a port and channel combination
+    fn get_channel_escrow_address(
```

Comment:
We should add a few tests to make sure that we generate the same address as `ibc-go` given the same input port/channel.

Reply:
For example, [this](https://www.mintscan.io/cosmos/account/cosmos1x54ltnyg88k0ejmk8ytwrhd3ltm84xehrnlslf) is the account on cosmos hub between cosmos hub and osmosis (channel `channel-141` and port `transfer`). We should make sure that we always get this address out given these inputs (and probably a few others).

Reply:
0d349d23d9a024e2d3766e3b86fff939fa917479